File: write-only-state.hs
module Main where
  
newtype Writer w a = Writer (a, w)

runWriter :: Writer w a -> (a, w)
runWriter (Writer (a, w)) = (a, w)


-- Functor
instance Functor (Writer w) where
  fmap f (Writer (a, w)) = Writer (f a, w)

-- Applicative
instance (Monoid w) => Applicative (Writer w) where
  pure a = Writer (a, mempty)
  (Writer (f, w1)) <*> (Writer (a, w2)) =
      Writer (f a, w1 `mappend` w2)

instance (Monoid w) => Monad (Writer w) where
  (Writer (a, w)) >>= k =
      let (a', w') = runWriter (k a)
      in Writer (a', w `mappend` w')

  return a = Writer (a, mempty)


-- Удобная функция для логирования
tell :: w -> Writer w ()
tell msg = Writer ((), msg)

-- Вычисление с логом
add :: Int -> Int -> Writer [String] Int
add x y = do
  tell ["Adding " ++ show x ++ " and " ++ show y]
  return (x + y)

mul :: Int -> Int -> Writer [String] Int
mul x y = do
  tell ["Multiplying " ++ show x ++ " and " ++ show y]
  return (x * y)

calc :: Writer [String] Int
calc = do
  a <- add 2 3
  b <- mul a 4
  return b

main :: IO ()
main = do
  let (result, log) = runWriter calc
  putStrLn $ "Result: " ++ show result
  putStrLn "Log:"
  mapM_ putStrLn log

File: write-only-state.ts
// Тип Writer: просто объект с результатом и логом
type Writer<W, A> = {
  value: A
  log: W
}

function writer<W, A>(value: A, log: W): Writer<W, A> {
  return { value, log }
}

function runWriter<W, A>(w: Writer<W, A>): [A, W] {
  return [w.value, w.log]
}

function fmap<W, A, B>(f: (a: A) => B, w: Writer<W, A>): Writer<W, B> {
  return writer(f(w.value), w.log)
}

function pure<W>(mempty: W) {
  return function <A>(a: A): Writer<W, A> {
    return writer(a, mempty)
  }
}

function apply<W>(
  mappend: (w1: W, w2: W) => W,
  wf: Writer<W, (a: any) => any>,
  wa: Writer<W, any>
): Writer<W, any> {
  return writer(wf.value(wa.value), mappend(wf.log, wa.log))
}

function bind<W, A, B>(
  mappend: (w1: W, w2: W) => W,
  wa: Writer<W, A>,
  k: (a: A) => Writer<W, B>
): Writer<W, B> {
  const wb = k(wa.value)
  return writer(wb.value, mappend(wa.log, wb.log))
}

function tell<W>(msg: W, mempty: W): Writer<W, void> {
  return writer(undefined, msg)
}

const mempty: string[] = []
const mappend = (w1: string[], w2: string[]) => w1.concat(w2)

function add(x: number, y: number): Writer<string[], number> {
  return writer(x + y, [`Adding ${x} and ${y}`])
}

function mul(x: number, y: number): Writer<string[], number> {
  return writer(x * y, [`Multiplying ${x} and ${y}`])
}

function calc(): Writer<string[], number> {
  return bind(mappend, add(2, 3), a =>
    bind(mappend, mul(a, 4), b =>
      bind(mappend, mul(b, b), c => 
        pure<string[]>(mempty)(c)
      )
    )
  )
}

function main() {
  const [result, log] = runWriter(calc())
  console.log("Result:", result)
  console.log("Log:")
  log.forEach(line => console.log(line))
}

main()
